### ISSUE TITLE
[FR] - cardano-cli - Make networkId configurable via environment variable like the socket path

### ISSUE THREAD

**Issue by ch1bo** [IntersectMBO/cardano-node#4418]
**Describe the feature you'd like**
The cardano-cli needs to know two things to interact with a `cardano-node`:
  
  1. Where to find the `cardano-node` socket file -> needs to be specified via environment variable `CARDANO_NODE_SOCKET_PATH`
  2. Which type of network to expect, either `mainnet` or `testnet` with a given `magic` -> needs to be specified as either `--mainnet` or `--testnet-magic XX`
  
When using the `cardano-cli` in scripting situations, it's quite annoying to mix both. Especially as the `--testnet-magic` option needs to be somewhat sensibly ordered in the command argument list.
Also, if both would be configurable via environment settings, we could provide shell environments where `cardano-cli query tip` would JustWork(tm).

Hence, I would like to do configure the "Network ID" (network discrimant + magic) via a `NETWORK_ID=[mainnet|testnet-XX]` environment variable as well.

**Describe alternatives you've considered**
Mixing setting an environment variable and interleaving `--testnet-magic ${magic}` arguments in scripts using `cardano-cli`.

Ideally, the `cardano-cli` would know about all "known" test networks already and we could do `mainnet|preprod|preview|custom-XX`, but this is more wishful thinking and maybe hard to maintain.

**Additional context / screenshots**
No

I would be able to implement this if we agree on an environment variable key + value syntax and precedence with command line options.

My suggestion for requirements:
* Make `--mainnet` and `--testnet-magic` in principle optional arguments
* Lookup `NETWORK_ID` environment variable on relevant `cardano-cli` commands
* If the env var is NOT SET and command line option NOT GIVEN -> fail like before when no cli options given
* If env var is SET and command line option is GIVEN -> cli options takes precedence
* If env var is SET and command line option NOT GIVEN -> this works now!

**Comment by rdlrt** [comment]
Building on #4402 , wouldn't it be nicer if CLI does not require magic/network ID at all, and just use network info via socket?

For the bit where network ID is required, eg: offline actions like creating address needs to know testnet vs mainnet,  the flags could be adjusted

**Comment by erikd** [comment]
@ch1bo My understanding is that the network team and the node team are already discussing this.

**Comment by gitmachtl** [comment]
I don't think that it should be done via an environment variable. Harder to run systems in parallel, would need an extra isolation.
As an additional input option it may be an option. I also created this Issue a while ago, that we should remove the magicId for testnets in the command, just --mainnet (should be the default) or optionally --testnet.

https://github.com/input-output-hk/cardano-node/issues/4409

**Comment by MarcFontaine** [comment]
I plan to implement the environment lookup as described in the initial feature request.
The plan is to add the new feature in a way that keeps `cardano-cli`  backwards compatible with earlier versions.
More preciseley:

* When the cli flags are given, they take precedence over `CARDANO_NETWORK_ID` and everything keeps working as before.
* When `CARDANO_NETWORK_ID` is set the flags can be left out.
* When neither `CARDANO_NETWORK_ID` is set nor flags are given, there will be an error.

However there is a catch ! One must take care of three special cases:
There are three command, where the network ID flag is already optional, and which use an ad hoc convention for the case when no network ID is specified:
* `cardano-cli transaction sign` -> defaults to `Byron`.
* `cardano-cli transaction witness` -> defaults to `Byron`
* `cardano-cli transaction calculate-min-fee` -> defaults to `mainnet`.

**Comment by gitmachtl** [comment]
so, the CARDANO_NETWORK_ID than holds a string like `--mainnet` and `-testnet-magic 1` as a 1:1 substitution to the current flags?

**Comment by ch1bo** [comment]
> string like --mainnet and --testnet-magic 1
I would have specified it to be `NETWORK_ID=mainnet` or `NETWORK_ID=testnet-1` respectively.

> its good to keep the flags
Seems like we have consensus that we need both, command line options *and* environment variables.

**Comment by gitmachtl** [comment]
> I would have specified it to be `NETWORK_ID=mainnet` or `NETWORK_ID=testnet-1` respectively.

Please do it with the `CARDANO_` prefix, so it is in line with the `CARDANO_NODE_SOCKET_PATH`, and grouped for this application in envs.

**Comment by CarlosLopezDeLara** [comment]
Resolved in https://github.com/IntersectMBO/cardano-node/pull/5119
